src/measurement/position_measurement.js

```diff
@@ -457,23 +457,21 @@ function coordsCharInner(cm, lineObj, lineNo, x, y) {
       ;({begin, end} = wrappedLineExtent(cm, lineObj, preparedMeasure, y))
     }
     pos = new Pos(lineNo, begin)
-    let beginLeft = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left
-    let dir = beginLeft < x ? 1 : -1
-    let prevDiff, diff = beginLeft - x, prevPos
-    do {
-      prevDiff = diff
+    let lastX = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left
+    let prevPos
+    let step = Math.abs(begin - end)
+    outer: do {
+      step = Math.ceil(step / 2)
       prevPos = pos
-      pos = moveVisually(cm, lineObj, pos, dir)
-      if (pos == null || pos.ch < begin || end <= (pos.sticky == "before" ? pos.ch - 1 : pos.ch)) {
-        pos = prevPos
-        break
+      for (let i = 0; i < step; ++i) {
+        let newPos = moveVisually(cm, lineObj, pos, lastX < x ? 1 : -1)
+        if (newPos == null || newPos.ch < begin || end <= (newPos.sticky == "before" ? newPos.ch - 1 : newPos.ch)) break outer
+        pos = newPos
       }
-      diff = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left - x
-    } while ((dir < 0) != (diff < 0) && (Math.abs(diff) <= Math.abs(prevDiff)))
-    if (Math.abs(diff) > Math.abs(prevDiff)) {
-      if ((diff < 0) == (prevDiff < 0)) throw new Error("Broke out of infinite loop in coordsCharInner")
-      pos = prevPos
-    }
+      lastX = cursorCoords(cm, pos, "line", lineObj, preparedMeasure).left
+      let adjPos = moveVisually(cm, lineObj, prevPos, 1)
+      if (adjPos && adjPos.ch == pos.ch) break
```

Comment:
If I understand these two lines right, they are for breaking out of the infinite loop alternating between the two cursor positions closest to the coordinates. I would prefer an approach that makes this intent obvious, instead of a side-effect. I also think this implementation is not correct for ltr spans, see the test I just pushed as d56a5f95b009f4cfa0ee0b4b4b4710348e34728b.